rmw_connextdds_common/include/rmw_connextdds/rmw_waitset_std.hpp

```diff
@@ -133,6 +128,31 @@ class RMW_Connext_Condition
     }
     return RMW_RET_OK;
   }
+
+  void
+  notify_waitset()
+  {
+    // We don't protect access to `waitset_mutex` with `mutex_internal` because
+    // we could still potentially miss the situation where the condition is
+    // being attached to the waitset (i.e. the application is inside
+    // WaitSet::wait(), possibly owning `waitset_mutex`) but `waitset_mutex`
+    // has not been set yet, in which case this notification will not be
+    // received by the waitset.
+    std::mutex * waitset_mutex = this->waitset_mutex;
+
+    if (nullptr != waitset_mutex) {
+      // Always take the WaitSet's mutex first to prevent deadlocks with
+      // concurrent WaitSet::wait().
+      std::lock_guard<std::mutex> lock_mutex(*waitset_mutex);
+      std::lock_guard<std::mutex> lock(this->mutex_internal);
+      if (nullptr != this->waitset_condition) {
+        this->waitset_condition->notify_one();
+      }
+    }
```

Comment:
I thought the problem was that we must guarantee that `std::condition_variable::notify_*()` is called only while the thread performing `WaitSet::wait()` is inside `std::condition_variable::wait()`, otherwise the notification will be lost. Hence why we must synchronize the call to `notify_one()` with the same mutex used (and released) by `condition_variable::wait()`.

Did I misunderstand the problem?

Reply:
It's not exactly that.

`cv.notify_one()` will do something if someone already called `cv.wait(mutex)` (i.e. a condition variable doesn't have memory).
To avoid losing notifications (and also, the spurios awakes that also `wait()` suffer), an external condition must be used and that condition must be modified with the cv mutex taken:

```cpp
// code notifying
{
  std::lock_guard<std::mutex> guard{cv_mutex};
  cv_flag = true;
}
cv.notify_one();

// other code resetting cv_flag
{
  std::lock_guard<std::mutex> guard{cv_mutex};
  cv_flag = false;
}

// code waiting on the condition
{
  std::unique_lock<std::mutex> guard{cv_mutex};
  while (!cv_flag) {
    cv.wait(mutex);
 }
 // or alternatively
 // cv.wait(mutex, [&cv_flag] (return cv_flag;));
}
```

doing that, you cannot lose modification.
If you were about to enter a `wait()`, two things can happen:

- the flag is set to `true` before taking the cv_mutex, `cv.wait()` will never be called.
- The flag cannot be set to true in the middle of checking the condition and calling `cv.wait()` becuase the `cv_mutex` is taken. The only other possibility is that the flag is set when `cv_wait()` internally released the mutex.
  You can either call `notify_one()` with the mutex taken or not.
  Releasing it before is an "optimization", to avoid the notified thread to awake and contend with the mutex again.

---

This case is more complex, as we have multiple conditions.
The only thing to be careful is to make sure to always take the `cv_mutex` before modifying each of the conditions.

Reply:
Thank you for clarifying, hopefully [b35d211](https://github.com/ros2/rmw_connextdds/pull/22/commits/b35d2114ef3b072f7ba386d1406eff89756d2288) resolves this.